cuML/src/decisiontree/kernels/gini.cuh

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <utils.h>
+#include "cub/cub.cuh"
+#include "../memory.cuh"
+#include <vector>
+#include "gini_def.h"
+
+template<>
+float set_min_val() {
+	return FLT_MAX;
+}
+
+template<>
+double set_min_val() {
+	return DBL_MAX;
+}
+
+
+template<class T>
+void GiniQuestion<T>::set_question_fields(int cfg_bootcolumn, int cfg_column, int cfg_batch_id, int cfg_nbins, int cfg_ncols, T cfg_min, T cfg_max, T cfg_value) {
+	bootstrapped_column = cfg_bootcolumn;
+	original_column = cfg_column;
+	batch_id = cfg_batch_id;
+	min = cfg_min;
+	max = cfg_max;
+	nbins = cfg_nbins;
+	ncols = cfg_ncols;
+	value = cfg_value; // Will be updated in make_split
+}
+
+__global__ void gini_kernel(const int* __restrict__ labels, const int nrows, const int nmax, int* histout)
+{
+	int tid = threadIdx.x + blockIdx.x * blockDim.x;
+	extern __shared__ unsigned int shmemhist[];
+	if (threadIdx.x < nmax)
+		shmemhist[threadIdx.x] = 0;
+
+	__syncthreads();
+
+	if (tid < nrows) {
+		int label = labels[tid];
+		atomicAdd(&shmemhist[label], 1);
+	}
+
+	__syncthreads();
+
+	if (threadIdx.x < nmax)
+		atomicAdd(&histout[threadIdx.x], shmemhist[threadIdx.x]);
+
+	return;
+}
+
+template<typename T>
+void gini(int *labels_in, const int nrows, const TemporaryMemory<T> * tempmem, GiniInfo & split_info, int & unique_labels, const cudaStream_t stream = 0)
```

Comment:
#257 and my PR #455 is an effort to remove all the default stream parameters. Please consider removing default stream = 0 in this implementation. It seems that you are not using the `const cudaStream_t stream = 0` in this function anyway :)

Reply:
thanks @chirayuG-nvidia  for spotting, i will remove that. its not needed here

Reply:
Thank you @chirayuG-nvidia . We removed default stream in a couple other places too in the most recent commit.